0.1.0 dev

mwbenowitz opened the pull request:
This is a pre-release version of the boilerplate code that includes readability, documentation and flexibility updates.

- Docstrings have been added for all methods within the code
- Certain methods have been moved from the `scripts` file to the various `helper` modules, which is a more logical place for their functionality
- Made the event source mapping handler more format agnostic. This merely accepts the JSON fields defined by the user, which is necessary to handle the variety of configuration options available. AWS will return logical error messages if an event source mapping is misconfigured
- Updated the `pyyaml` loader to address a security concern from the library
- Updated the README file

mwbenowitz committed:
Refactor sript/helper code for better organization

Moves various helper functions from the `scripts` file to
various `helper` classes, which is a more logical organization
for this repo, dividing supporting code from the method in which
it is invoked.

Further the test suite has been updated to reflect this new
structure

mwbenowitz committed:
Fix pyyaml security issue with `full_load`

mwbenowitz committed:
Add docstrings and other documentation updates

This adds docstrings describing the general operation of the
generic function implementation.

This also removes some code from the `handler()` method that
was specific to `Kinesis` invocation events, which is too specific
for the intended uses of this boilerplate function

mwbenowitz committed:
Update README and bump version

mwbenowitz committed:
Add status badge to README

mwbenowitz committed:
Replace event source mapping with dict comprehension

Previously the event source mapping objects were being built with
explicit dictionary keys, which made it difficult to manage different
event source types. Switching to a dict comprehension allows
different sources to be easily handled.

The only drawback to this is that it relies on the user to provide a
properly formatted event source mapping. However, if a faulty event
source mapping is used the deployment process will fail with a fairly
descriptive error messsage that should allow the error to be easily
fixed.

mwbenowitz committed:
Lint files and add linting to CI tests

seanredmond commented on helpers/configHelpers.py:
```diff
@@ -28,5 +43,55 @@ def loadEnvFile(runType, fileString):
+    # Overwrite/add any vars in the core config.yaml file
+    configDict = loadEnvFile(runType, None)
+
+    combinedConfig = ChainMap(envDict, configDict)
```
If all you're doing is combining two dicts, this is faster (in versions 3.5+) as long as it's okay for values in the second dict to overwrite matching keys in the first.:

    {**envDict, **configDict}

I think ChainMap does the opposite when you flatten one to a dict. That, is it keeps values from the first. So the equivalent of:

    dict(ChainMap(envDict, configDict))

may be:

    {**configDict, **envDict}

mwbenowitz replied:
Thanks for the tip, I wasn't aware of that construction. And yes, the idea is explicitly to have the second dict overwrite values in the first (so you can create env-specific variables for production that overwrite the general settings used in local/production/qa branches, etc.)

seanredmond commented on scripts/lambdaRun.py:
```diff
@@ -2,27 +2,23 @@
+        InvalidExecutionType: If the args do not contain a valid execution type
+        raise an error.
+    """
     if len(sys.argv) != 2:
```
Is it more important that there is at least one? E.g.:

    try:
        runType = sys.argv[1]
    except IndexError:
        logger.warning('Missing argument')
        sys.exit(1)

mwbenowitz replied:
likely yes? I'm unsure what's the best practice here. Additional `args` will have no effect, but should that be a valid execution type? I'm torn between being exacting in how this is invoked and tolerating additional parameters that will have no effect either way. Perhaps it should raise an `info` message if more than one arg is supplied and still execute?

seanredmond replied:
So what could happen? Maybe through a copy-paste error you end up with something like

    scripts.lambdaRun development build-development

Where you mean for the second thing to run. If that is potentially really bad, and/or hard to track down, maybe it _is_ better to check that there is only one argument?

mwbenowitz replied:
The worst thing that could happen is that you accidentally deploy or deploy to the wrong environment. Of course you can do that anyway, but if you hit `make deploy production` when you wanted `make deploy development` I'm not sure there's much we can do to help. (Perhaps a future feature should prompt the user to confirm production deployments, or supply a flag to not require that).

Hypothetically local testing could be destructive too, but if your local code is integrated to a feature in production, again, not much we can do.

My idea was that more than one argument is not intended and should be flagged because we have no way of knowing what the intended command was. 

Also, perhaps I'm not thinking this through, but I don't think this should ever be directly invoked by a user, it should only handle calls from the `Makefile`, though I guess we should support people building different commands that they want to use?

seanredmond replied:
Maybe just leave it as is.

seanredmond commented on scripts/lambdaRun.py:
```diff
@@ -57,7 +53,9 @@ def main():
 
     elif re.match(r'^build-(?:development|qa|production)', runType):
```
the if/elif plus the regexes seems a little multi-dimensional and hard to debug if there's a subtle problem in a regex. With one exception, the same code is getting run for each, just with some different variables. I can think of ways to do it with a dict where the values are functions to run, then with something like:

    funcMap.get(runType, errFunc)()

You can run the function matching the runType, or a function to log an error if nothing matches. The advantage would be that the mapping would contain all the explicitly valid runTypes and you don't have to pick your way through elifs and regexes.

mwbenowitz replied:
That's an excellent idea. I haven't really thought too much about this function/file since the beginning because it hasn't posed any issues, but that does seem like a DRY-er and cleaner implementation. anything that eliminates regexes is generally 👍 in my book.

Will take a stab at implementing something along these lines.

mwbenowitz committed:
Replace `ChainMap` with faster `dict` merging

Implenting the merging of two configuration dictionaries with
`ChainMap` proved to be unnecessary/slower. This has been removed,
with the replacement tested.

Revealed in this testing is that neither the merge nor `ChainMap`
will update nested `dict`s, merely replace them wholesale. This
means that configuration settings, particularly environment
variables must be duplicated wholesale across environments, not
only those that change. This should be an area of improvement
in the future.

mwbenowitz committed:
Refactor `lambdaRun` for better maintainability

This refactors the `regex` based structure of the file that
managed the specific type of invocation being made to run the
lambda function.

This is replaced with an explicit mapping between the valid
execution commands and the methods that they invoke. In the case
that a command is received that is not found and error method
is called, which raises an `InvalidExectutionType` error.

This extracts a large amount of functionality from a single method
into several smaller methods, allowing for easier tracing of issues
and clearer visualisation of the path through this module.

mwbenowitz committed:
Comment typos and clarifications

mwbenowitz committed:
Remove extraneous print statement

mwbenowitz committed:
Linting whitespace fixes
